### ISSUE TITLE
Quokka constantly glitches out and shows results offset heavily

### ISSUE THREAD

**Issue by rsxdalv** [wallabyjs/quokka#375]
### Sample code

```javascript
console.log('...'); 3
```

### Sample repository link

*If the issue can not be reproduced just using the quokka file above 
(for example because it requires/imports some files from your project), 
please create a small repository where the issue can be reproduced.* 

### Quokka.js Console Output

-- the console is empty --

### Code editor version

Version: 1.34.0 (user setup)
Commit: a622c65b2c713c890fcf4fbf07cf34049d5fe758
Date: 2019-05-15T21:59:37.030Z
Electron: 3.1.8
Chrome: 66.0.3359.181
Node.js: 10.2.0
V8: 6.6.346.32
OS: Windows_NT x64 10.0.17763

### OS name and version

Windows (no WSL)

Clue - this is probably related to pushing the memory limits, but once it happens, the only way to get any output is to log about a 100 lines till the cutoff ends.

Occasionally on errors I get a very long console output, which I just now was 285143 lines, sample is here:
```

[ '1a', '2b', '3c', '4d', '5e' ] 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
 
[ '1a', '2b', '3c', '4d', '5e' ] 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​


[ '1a', '2b', '3c', '4d', '5e' ] 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
 
[ '1a', '2b', '3c', '4d', '5e' ] 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
 
confirm is not defined 
  at ​​​Object.<anonymous>​​​ ​./quokka.js:83​
  at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
 
cponm is not defined 
  at ​​​Object.<anonymous>​​​ ​./quokka.js:83​
  at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
 
c is not defined 
  at ​​​Object.<anonymous>​​​ ​./quokka.js:83​
  at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
 
conso is not defined 
  at ​​​Object.<anonymous>​​​ ​./quokka.js:83​
  at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​

cmon ​​​​​at ​quokka.js:1:1​
 
[ '1a', '2b', '3c', '4d', '5e' ] 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​

cmon ​​​​​at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​

cmon ​​​​​at ​quokka.js:1:2​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​

cmon ​​​​​at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
 
w is not defined 
  at ​​​Object.<anonymous>​​​ ​./quokka.js:5​
  at ​quokka.js:1:1​

cmon ​​​​​at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​

cmon ​​​​​at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​

... ​​​​​at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
Failed to instrument quokka.js 
> 1 |  console.log('...'); 3l; 
    |                       ^ SyntaxError: Identifier directly after number (1:22) 
  2 | // const R = require('ramda'); 
  3 |  
  4 | // const fragmentRef = { 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​

... ​​​​​at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
Failed to instrument quokka.js 
> 1 |  console.log('...'); 3 231 
    |                        ^ SyntaxError: Unexpected token (1:23) 
  2 | // const R = require('ramda'); 
  3 |  
  4 | // const fragmentRef = { 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
Failed to instrument quokka.js 
> 1 |  console.log('...'); 3 2314 
    |                        ^ SyntaxError: Unexpected token (1:23) 
  2 | // const R = require('ramda'); 
  3 |  
  4 | // const fragmentRef = { 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
Failed to instrument quokka.js 
> 1 |  console.log('...'); 3 2314324 
    |                        ^ SyntaxError: Unexpected token (1:23) 
  2 | // const R = require('ramda'); 
  3 |  
  4 | // const fragmentRef = { 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​

... ​​​​​at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
Failed to instrument quokka.js 
> 1 |  console.log('...'); 3[] 
    |                        ^ SyntaxError: Unexpected token (1:23) 
  2 | // const R = require('ramda'); 
  3 |  
  4 | // const fragmentRef = { 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​

... ​​​​​at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
 
Cannot read property '4' of undefined 
  at ​quokka.js:1:1​
  at ​​​Object.<anonymous>​​​ ​./quokka.js:5​

... ​​​​​at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
Failed to instrument quokka.js 
> 1 |  console.log('...'); 3[3][4]2 
    |                             ^ SyntaxError: Unexpected token (1:28) 
  2 | // const R = require('ramda'); 
  3 |  
  4 | // const fragmentRef = { 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
 
Cannot read property '4' of undefined 
  at ​quokka.js:1:1​
  at ​​​Object.<anonymous>​​​ ​./quokka.js:5​

... ​​​​​at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
Failed to instrument quokka.js 
> 1 |  console.log('...'); 3[3][4]2 
    |                             ^ SyntaxError: Unexpected token (1:28) 
  2 | // const R = require('ramda'); 
  3 |  
  4 | // const fragmentRef = { 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
Failed to instrument quokka.js 
> 1 |  console.log('...'); 3[3][4]23 
    |                             ^ SyntaxError: Unexpected token (1:28) 
  2 | // const R = require('ramda'); 
  3 |  
  4 | // const fragmentRef = { 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
Failed to instrument quokka.js 
> 1 |  console.log('...'); 3[3][4]2 
    |                             ^ SyntaxError: Unexpected token (1:28) 
  2 | // const R = require('ramda'); 
  3 |  
  4 | // const fragmentRef = { 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
 
Cannot read property '4' of undefined 
  at ​quokka.js:1:1​
  at ​​​Object.<anonymous>​​​ ​./quokka.js:5​

... ​​​​​at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
Failed to instrument quokka.js 
> 1 |  console.log('...'); 3[3][4]2 
    |                             ^ SyntaxError: Unexpected token (1:28) 
  2 | // const R = require('ramda'); 
  3 |  
  4 | // const fragmentRef = { 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
Failed to instrument quokka.js 
  95 | //     fragmentRef 
  96 | //     ) 
> 97 |  
     | ^ SyntaxError: Unexpected token (97:0) 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​

... ​​​​​at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
Failed to instrument quokka.js 
> 1 |  console.log('...'); 3[3]4 
    |                          ^ SyntaxError: Unexpected token (1:25) 
  2 | // const R = require('ramda'); 
  3 |  
  4 | // const fragmentRef = { 
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​

... ​​​​​at ​quokka.js:1:1​
​​​​Quokka 'Untitled-3.js' (node: v10.15.3)​​​​
Failed to instrument quokka.js 
> 1 |  console.log('...'); 3[3]1 
    |                          ^ SyntaxError: Unexpected token (1:25) 
  2 | // const R = require('ramda'); 
  3 |  
  4 | // const fragmentRef = { 
```

**Comment by rsxdalv** [comment]
Nevermind, this happens even with tiny amounts of data.
I have a curly bracket colorizer that manages to colorize quokka output as well, could that be the issue?

**Comment by rsxdalv** [comment]
It's something about this string 's\u0000\u0011'(...) being logged to the console

**Comment by rsxdalv** [comment]
Nevermind, it just seems to happen constantly almost whatever I do.

**Comment by smcenlly** [comment]
We've had some problems with `bracket colorizers` in the past. For example, see #327.

Can you please confirm the following:
1) Does Quokka work without the bracket colorizer installed?
2) Which bracket colorizer extension are you using?
3) What steps do you take to reproduce the issue?

From what we saw in the past, some of the bracket colorizer plugins perform some async processing and override the default text selection so that may be part of the cause.

**Comment by spryce** [comment]
I bought a couple of pro licenses a few months back and today figured I would give it another go. 

1. Does Quokka work without the bracket colorizer installed?
I removed my bracket colorizer plug in after reading this comment and that definitely helped. 
Unfortunately after reaching 20 lines of code it just stops displaying any inline feedback other than undefined messages. If I add the //? to vars it will display results in the output window but usually only after I save. It's slow to update, or just doesn't update.

2. Which bracket colorizer extension are you using?
Rainbow Brackets - removed and restarted.

3. What steps do you take to reproduce the issue?
Just general experimenting unfortunately.

![quokka](https://user-images.githubusercontent.com/12742003/58610390-aeeb6300-82ee-11e9-99ba-9f7cec337c1d.png)
![quokka-1](https://user-images.githubusercontent.com/12742003/58610391-aeeb6300-82ee-11e9-879b-cecdf8fa1ad0.png)
![quokka-2](https://user-images.githubusercontent.com/12742003/58610392-aeeb6300-82ee-11e9-8771-979ca2d727b2.png)

The last image here shows an undefined for te even though test is in this case is actually printing inline. 

VS Code
Version: 1.34.0 (user setup)
Commit: a622c65b2c713c890fcf4fbf07cf34049d5fe758
Date: 2019-05-15T21:59:37.030Z
Electron: 3.1.8
Chrome: 66.0.3359.181
Node.js: 10.2.0
V8: 6.6.346.32
OS: Windows_NT x64 10.0.17763

**Comment by ArtemGovorov** [comment]
@spryce 
- What exact command are you using to start Quokka on your file?
- What other extensions do you have installed?

**Comment by spryce** [comment]
@ArtemGovorov 
I played with several commands on existing files and new but the screens were taken after doing:

- New JavaScript File 
- Start on Current File (after re-opening the editor etc)

Extensions:

- ESLint 1.9.0 
- Debugger for Chrome 4.11.3.
- Quokka 1.0.219

**Comment by rsxdalv** [comment]
Here's an example (colorizer still enabled) where an error causes messages to disappear.

![image](https://user-images.githubusercontent.com/6757283/58619109-d6c9cf00-82cc-11e9-9af0-bb76ec538853.png)

Same goes for /*?*/ syntax, it only works when forced with /*? $*/, but in iterative code (haven't had the time to create a minimal example yet) it has a tendency to only display first value (even when no errors are thrown).

![image](https://user-images.githubusercontent.com/6757283/58619224-23150f00-82cd-11e9-9208-5a80fec2203c.png)

Now this happened
![image](https://user-images.githubusercontent.com/6757283/58619257-3d4eed00-82cd-11e9-8e16-474328e154bb.png)

Current list of extensions:
```
2gua.rainbow-brackets
alefragnani.Bookmarks
Damien.autoit
dcealopez.pawn-scripting-language-support
donjayamanne.githistory
DotJoshJohnson.xml
eamodio.gitlens
eamodio.toggle-excluded-files
emilast.LogFileHighlighter
fabiospampinato.vscode-todo-plus
formulahendry.code-runner
geddski.macros
jingkaizhao.vscode-redux-devtools
jkjustjoshing.vscode-text-pastry
jock.svg
jpoissonnier.vscode-styled-components
JuanCasanova.awesometypescriptproblemmatcher
karunamurti.haml
Kasik96.swift
krizzdewizz.refactorix
MariusAlchimavicius.json-to-ts
michelemelluso.code-beautifier
mjmcloug.vscode-elixir
ms-vscode.csharp
ms-vscode.Go
msjsdiag.debugger-for-chrome
rebornix.ruby
rebornix.toggle
robinbentley.sass-indented
rsxlv.rsxlv-personal
ryu1kn.partial-diff
sensourceinc.vscode-sql-beautify
Shan.code-settings-sync
spywhere.express-lines
stringham.move-ts
techer.open-in-browser
trixnz.vscode-lua
vscodevim.vim
WallabyJs.quokka-vscode
wmaurer.change-case
```

Next, I've disabled rainbow brackets

Still managed to score this error with just copying console.log (this is minor, but a new one I've noticed):
![image](https://user-images.githubusercontent.com/6757283/58619585-d7169a00-82cd-11e9-96cd-aa55cd4d4fc2.png)

Same behavior here (only first 2 are logged until /*? $*/ workaround is used)
![image](https://user-images.githubusercontent.com/6757283/58619643-f57c9580-82cd-11e9-99f6-cd72228b424c.png)

![image](https://user-images.githubusercontent.com/6757283/58619683-0a592900-82ce-11e9-9ae0-e7a6a337a2c5.png)

I haven't been able to reproduce the bug right now (I'll let you know once my next real session of development comes around and it'll probably come back), but it starts to seem connected to a part of this bag of things: clearing the output window, large strings, { recycle: true }, colorized util.inspect.

I did manage to score this, however:
```
{ [32m'0'[39m: [33m80821078[39m, 
  [32m'1'[39m: [ [33m4.696337699890137[39m, [33m1.3024797439575195[39m, [length]: [33m2[39m ], 
  [32m'2'[39m: [33m254.92654418945312[39m, 
  [32m'3'[39m: [ [33m13.091056823730469[39m, [33m-1.193798542022705[39m, [length]: [33m2[39m ], 
  [32m'4'[39m: [33m7.699999809265137[39m, 
  [32m'253'[39m: [33m18[39m } 
```

using this:
```
const ins2 = myObject =>
    console.log(util.inspect(myObject, false, null, true /* enable colors */))
```

**Comment by rsxdalv** [comment]
To add: now I'm again experiencing old errors from a different file flowing into my current file (note, I'm working on Untitled-5 not slicer.js):
![image](https://user-images.githubusercontent.com/6757283/58653782-220ccd80-831f-11e9-9e48-d6ee7dbf731b.png)

I can only think that this is because I'm smashing too much data into Quokka, until something has failed internally.

**Comment by smcenlly** [comment]
Could you please try disabling the `vim` extension and seeing if that fixes your problem?

We have previously had trouble with the `vim` extension (due to its behavior, not Quokka). Please see https://github.com/wallabyjs/quokka/issues/326 (and below). The issue may not be the same, but if removing the `vim` extension fixes your problem, then I would suggest you raise this issue with the `vim` team. If different from what we previously identified, we are happy to assist in investigating the root cause but fixing the problem is likely outside our area of expertise.

> We've done some debugging and this issue is caused by the Vim extension, not Quokka.
>
> What's happening is a keystroke occurs and is processed by both Wallaby and the Vim extension. Vim processes the keyboard commands asynchronously. At the same time, Wallaby updates the Output window, which causes VS Code to temporarily refer to it as the vscode.window.activeTextEditor. When the Vim extension processes the keystroke, it always uses the activeTextEditor (which is now the Output window), instead of remembering the textEditor at the time of the event. After this point, the Vim extension's mechanism for remembering the cursor position is messed up until it's reset by a manual shift in focus (e.g. I-Beam cursor click in the text editor).
>
> I'm not entirely sure whether this is also a problem in VS Code (not sure that the activeTextEditor should change at all while we're updating the console). Having had a look at the use of activeTextEditor in the Vim plugin, I would have thought the activeTextEditor should be passed at the time the command is processed.
>
> I would suggest that you report this and provide these details to someone on the Vim team to investigate/fix.

**Comment by rsxdalv** [comment]
Thanks for the input, I haven't had the chance to test it yet, but I have a question - does that apply if I almost never use VIM keybindings - is it an issue with the extension in background, or foreground, if you know?

**Comment by smcenlly** [comment]
I'm not sure as I don't use the extension. You will need to test it.

**Comment by smcenlly** [comment]
I'm going to close this issue for now. If you find that VIM is fixed and is now working with Quokka without issue, please let us know the conflicting plugin and we will help you investigate further.

**Comment by rsxdalv** [comment]
Disabled VIM
Missing logs on error still persists:
```js
const z = 5;

console.log(3);
console.log(3);
console.log(3);
console.log(3);
console.log(3);
console.log(3);
console.log(3);
console.log(3);
console.log(3);
console.log(3);
throw new Error(4);
console.log(3);
console.log(3);
console.log(3);
console.log(3);
console.log(3);
```

Output (same in blue text):
```
​​​​Quokka 'Untitled-1.js' (node: v10.15.3)​​​​
 
4 
  at ​quokka.js:13:0​
  at ​​​Object.<anonymous>​​​ ​./quokka.js:23​

3 ​​​​​at ​quokka.js:3:0​

3 ​​​​​at ​quokka.js:4:0​

3 ​​​​​at ​quokka.js:5:0​
```

**Comment by spryce** [comment]
Not sure how this issue is deemed resolved. I'm not using VIM

**Comment by ArtemGovorov** [comment]
@rsxdalv I have been able to reproduce the issue you've been reporting (with multiple console.log statements and only first few being logged if there's an error). We'll investigate it tomorrow.

@spryce I'm not sure if you actually have the same issue as the topic starter. Could you please provide a full source code of your `event-prototyping.js` file os we could try to reproduce the issue?

**Comment by smcenlly** [comment]
@rsxdalv - Thanks for your patience as we worked through this issue. We've now found and fixed the problem.

@spryce - we believe that this will also solve your issue. If it does not, please open a [new issue](https://github.com/wallabyjs/quokka/issues/new) with the `event-prototyping.js` source code and any relevant steps for duplicating the issue.

---

You will need to update Quokka to get this fix. The latest versions are:

* [Quokka for IntelliJ v1.0.173](https://plugins.jetbrains.com/plugin/9667-quokka)
* [Quokka for VSCode v1.0.223](https://marketplace.visualstudio.com/items?itemName=WallabyJs.quokka-vscode)
* [Quokka for Atom v1.0.150](https://libraries.io/atom/atom-quokka/1.0.150)

**Comment by rsxdalv** [comment]
Thank you for your time!
Verified that it works. Since I haven't encountered the original issues (nor have I dissected packets and massive buffers) for now, I think this should be it. I'll let you know if I manage to reproduce the scroll issue (and whether or not disabling rainbow brackets and VIM affects it)

**Comment by rsxdalv** [comment]
No, the original error still persists: https://www.youtube.com/watch?v=fGLLVB4hB78

**Comment by ArtemGovorov** [comment]
@rsxdalv This may be VS Code output channel glitch (I have seen something like it recently myself with other extension output). Try restarting VS Code to see if the issue persists.

**Comment by rsxdalv** [comment]
It didn't in the very first printout but the second I made changes it came back. I think it's caused by invalid characters being printed when working with binary data.

**Comment by ArtemGovorov** [comment]
@rsxdalv Could you please share some code to reproduce the issue?

**Comment by rsxdalv** [comment]
Comment/Uncomment some of the console.log lines to trigger it:
```js
const sources = {
    "coords": { "x": -808.0108032226562, "y": 1300.3184814453125, "z": 258.6042785644531, "h": 299.8853759765625 },
}

// const buf = new ArrayBuffer(8);
// const view = new DataView(buf);
const xs = [].concat(
    ...Object.keys(sources).map(key =>
        Object.keys(sources[key]).map(key2 =>
            sources[key][key2]
        )
    )
)
const buf = new ArrayBuffer(xs.length * 4);
const view = new DataView(buf);
xs.map((y, i) =>
    view.setFloat32(i * 4, y)
)
const viewBuf = new Buffer(buf);
xs.toString().length /*?*/
JSON.stringify(xs).length /*?*/
const utf8Encoded = viewBuf.toString('utf8');
const utf8Decoded = Buffer.from(utf8Encoded, 'utf8')
viewBuf.toString('hex') /*?*/
utf8Decoded.toString('hex') /*?*/
viewBuf /*?*/
utf8Decoded /*?*/
const byteCount = utf8Encoded.length /*?*/
console.log(
    (byteCount/363) ** -1,
    byteCount/363,
    // utf8Encoded,
    viewBuf.toString('hex'),
    utf8Decoded.toString('hex'),
)
```

**Comment by ArtemGovorov** [comment]
@rsxdalv Thanks! We'll investigate it.

**Comment by smcenlly** [comment]
Thanks for the sample. We've identified that your content was being truncated because the `viewBuf` and `utf8Decoded` strings contained the NUL character (`\u0000`). After encountering a NUL character in the output, the VS Code output channel no longer behaves consistently/correctly.

We've updated Quokka to remove NUL characters from the output, which will fix your problem. This is available in Quokka for VS Code v1.0.234. We've also raised this [an issue for the VS Code team to investigate/address](https://github.com/microsoft/vscode/issues/77377).
